File: BrickKit/app/src/main/java/com/wayfair/brickkitdemo/SimpleBrickFragment.java
/*
 * Copyright © 2017-2020 Wayfair. All rights reserved.
 */
package com.wayfair.brickkitdemo;

import android.os.Bundle;

import com.wayfair.brickkit.brick.BaseBrick;
import com.wayfair.brickkit.brick.ViewModelBrick;
import com.wayfair.brickkit.padding.BrickPaddingFactory;
import com.wayfair.brickkit.size.FullPhoneFullHalfTabletBrickSize;
import com.wayfair.brickkitdemo.datamodel.TextDataModel;
import com.wayfair.brickkitdemo.viewmodel.TextViewModel;

import java.util.LinkedList;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Example fragment which shows text bricks.
 * <p>
 * In portrait, the bricks are full width.
 * In landscape the bricks are half width.
 */
public class SimpleBrickFragment extends BrickFragment {
    private int numberOfBricks = 100;

    LinkedList<BaseBrick> bricks;

    /**
     * Create a new instance of a SimpleBrickFragment.
     *
     * @param numberOfBricks The number of bricks you want in the sub-{@link com.wayfair.brickkit.BrickDataManager}
     * @return The SimpleBrickFragment your created
     */
    public static SimpleBrickFragment newInstance(int numberOfBricks) {
        SimpleBrickFragment fragment = new SimpleBrickFragment();
        fragment.numberOfBricks = numberOfBricks;
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        BrickPaddingFactory brickPaddingFactory = new BrickPaddingFactory(getResources());
        bricks = new LinkedList<>();

        for (int i = 0; i < numberOfBricks; i++) {
            ViewModelBrick brick = genBrick(i, brickPaddingFactory);

            bricks.add(brick);
            dataManager.addLast(brick);
        }
    }

    /**
     * Generate a brick.
     *
     * @param i the index of that brick, also uses in the text of the brick
     * @param brickPaddingFactory {@link BrickPaddingFactory} to generate padding with
     * @return the brick
     */
    public ViewModelBrick genBrick(int i, BrickPaddingFactory brickPaddingFactory) {
        final TextDataModel dataModel = new TextDataModel("Brick: " + i);

        new Timer().schedule(new TimerTask() {
            @Override
            public void run() {
                dataModel.appendText(" Hello");
            }

        }, 0, 1000);

        return new ViewModelBrick.Builder(R.layout.text_brick_vm)
                .setPadding(brickPaddingFactory.getInnerOuterBrickPadding(R.dimen.four_dp, R.dimen.eight_dp))
                .setSpanSize(new FullPhoneFullHalfTabletBrickSize())
                .addViewModel(
                        BR.textViewModel,
                        new TextViewModel(dataModel)
                )
                .build();
    }
}

File: BrickKit/app/src/main/java/com/wayfair/brickkitdemo/simplebrick/SimpleBrickFragment.kt
/*
 * Copyright © 2017-2020 Wayfair. All rights reserved.
 */
package com.wayfair.brickkitdemo.simplebrick

import android.os.Bundle
import com.wayfair.brickkit.brick.ViewModelBrick
import com.wayfair.brickkit.padding.BrickPaddingFactory
import com.wayfair.brickkit.size.FullPhoneFullHalfTabletBrickSize
import com.wayfair.brickkitdemo.BR
import com.wayfair.brickkitdemo.BrickFragment
import com.wayfair.brickkitdemo.R
import com.wayfair.brickkitdemo.simplebrick.datamodel.TextDataModel
import com.wayfair.brickkitdemo.simplebrick.viewmodel.TextViewModel
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Example fragment which shows text bricks and updates each second.
 */
class SimpleBrickFragment : BrickFragment() {
    private var numberOfBricks = 100

    private val executor = ScheduledThreadPoolExecutor(1)
    private val dataModels by lazy { (0 until numberOfBricks).map { i -> TextDataModel("Brick: $i") } }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        dataModels.forEach { dataModel ->
            ViewModelBrick.Builder(R.layout.text_brick_vm)
                .setPadding(BrickPaddingFactory(resources).getInnerOuterBrickPadding(R.dimen.four_dp, R.dimen.eight_dp))
                .setSpanSize(FullPhoneFullHalfTabletBrickSize())
                .addViewModel(BR.viewModel, TextViewModel(dataModel))
                .build()
                .addLastTo(dataManager)
        }
    }

    override fun onResume() {
        super.onResume()

        dataModels.forEach { dataModel ->
            executor.scheduleAtFixedRate({ dataModel.appendText(" Hello") }, 0, 1, TimeUnit.SECONDS)
        }
    }

    override fun onPause() {
        super.onPause()

        executor.purge()
    }

    companion object {
        /**
         * Create a new instance of a SimpleBrickFragment.
         *
         * @param numberOfBricks The number of bricks you want in the fragment
         * @return The SimpleBrickFragment you created
         */
        fun newInstance(numberOfBricks: Int = 100) = SimpleBrickFragment().apply {
            this.numberOfBricks = numberOfBricks
        }
    }
}
